cupyx/cutensor.pyx

```diff
@@ -1066,3 +1078,682 @@ def _try_elementwise_binary_routine(
+# MgContractionDescriptor: This class encapsulates the opaque structure
+# `cutensorMgContractionDescriptor_t` representing a plan for multi-GPU
+# contraction.
+###############################################################################
+
+cdef class MgContractionDescriptor:
+    cdef intptr_t _ptr
+
+    def __init__(self, intptr_t handle, intptr_t descA, intptr_t modeA,
+                 intptr_t descB, intptr_t modeB, intptr_t descC,
+                 intptr_t modeC, intptr_t descD, intptr_t modeD,
+                 int compute):
+        self._ptr = cutensor.createMgContractionDescriptor(
+            handle, descA, modeA, descB, modeB, descC, modeC, descD,
+            modeD, compute)
+
+    def __dealloc__(self):
+        if self._ptr is not 0:
+            cutensor.destroyMgContractionDescriptor(self._ptr)
+        self._ptr = <intptr_t>NULL
+
+    @property
+    def ptr(self):
+        return self._ptr
+
+
+###############################################################################
+# MgContractionFind: This class encapsulates the opaque structure
+# `cutensorMgContractionFind_t` representing a find object to set
+# preferences for algorithm selection.
+###############################################################################
+
+cdef class MgContractionFind:
+    cdef intptr_t _ptr
+
+    def __init__(self, intptr_t handle, int algo):
+        self._ptr = cutensor.createMgContractionFind(handle, algo)
+
+    def __dealloc__(self):
+        if self._ptr is not 0:
+            cutensor.destroyMgContractionFind(self._ptr)
+        self._ptr = <intptr_t>NULL
+
+    @property
+    def ptr(self):
+        return self._ptr
+
+
+###############################################################################
+# MgContractionPlan: This class encapsulates the opaque structure
+# `cutensorMgContractionPlan_t` representing a plan for multi-GPU contraction.
+###############################################################################
+
+cdef class MgContractionPlan:
+    cdef intptr_t _ptr
+
+    def __init__(self, intptr_t handle, intptr_t desc, intptr_t find,
+                 intptr_t workspaceDeviceSize,
+                 uint64_t workspaceHostSize):
+        self._ptr = cutensor.createMgContractionPlan(
+            handle, desc, find, workspaceDeviceSize, workspaceHostSize)
+
+    def __dealloc__(self):
+        if self._ptr is not 0:
+            cutensor.destroyMgContractionPlan(self._ptr)
+        self._ptr = <intptr_t>NULL
+
+    @property
+    def ptr(self):
+        return self._ptr
+
+cpdef MgHandle _get_mg_handle(devices=None):
+    if devices is None:
+        devices = _numpy.arange(getDeviceCount(), dtype=_numpy.int32)
+    else:
+        devices = _numpy.asarray(sorted(devices), dtype=_numpy.int32)
+    key = tuple(devices)
+    if key not in _mg_handles:
+        _mg_handles[key] = MgHandle(devices)
+    return _mg_handles[key]
+
+
+class ndarray_mg:
```

Comment:
The interface of `DistributedArray` is different from `ndarray_mg`, so it may take very long time to replace. To prevent this, this pull request might continue with the existing `ndarray_mg` implementation, but could you mark it as an experimental feature in such case?

Reply:
I will try

Reply:
“but could you mark it as an experimental feature in such case?”
How can i do this? Add some note in the doc string is enough?

Reply:
Could you raise `FutureWarning` with `cupy._util.experimental`?
https://github.com/cupy/cupy/blob/2fd0b819bc00b5e0b16e86c1841a0d3176241593/cupy/_util.pyx#L80

Reply:
Done in commit [46dcbc2](https://github.com/cupy/cupy/pull/8212/commits/46dcbc209f3fbe9595a81a4d09bf26a01ea4d974)